File: src/com/github/stromland/piano/services/Scale.kt
package com.github.stromland.piano.services

object Scale {
    fun scales(): Map<String, List<Int>> {
        return mapOf(
            "major" to listOf(2, 2, 1, 2, 2, 2, 1),
            "minor" to listOf(2, 1, 2, 2, 1, 2, 2),
            "harmonicMinor" to listOf(2, 1, 2, 2, 1, 3, 1)
//            "pentatonicMajor" to listOf(2, 2, 3, 2, 3),
//            "pentatonicMinor" to listOf(3, 2, 2, 3, 2)
        )
    }

    fun getScaleKeyIndexes(noteIndex: Int, scaleName: String, sets: Int = 0): List<Int> {
        val scaleSpec = getScaleSpec(scaleName)
        val scale = (0..sets).fold(listOf<Int>()) { acc, _ -> acc + scaleSpec }

        return scale.fold(listOf()) { scaleKeys, nextKey ->
            when (scaleKeys.size) {
                0 -> listOf(noteIndex, noteIndex + nextKey)
                else -> scaleKeys + (scaleKeys.last() + nextKey)
            }
        }
    }

    private fun getScaleSpec(scaleName: String): List<Int> {
        return scales()[scaleName] ?: throw IllegalArgumentException("Not a valid scale scaleName ($scaleName)")
    }
}

File: packages/piano-react/src/services/Scale.ts
interface Scales {
  major: number[];
  minor: number[];
  harmonicMinor: number[];
  // pentatonicMajor: number[];
  // pentatonicMinor: number[];
}

export type ScaleType = keyof Scales;

export class Scale {
  public static scales: Scales = {
    major: [2, 2, 1, 2, 2, 2, 1],
    minor: [2, 1, 2, 2, 1, 2, 2],
    harmonicMinor: [2, 1, 2, 2, 1, 3, 1],
    // pentatonicMajor: [2, 2, 3, 2, 3],
    // pentatonicMinor: [3, 2, 2, 3, 2],
  };

  public static getScaleKeyIndexes(
    noteIndex: number,
    scaleType: ScaleType,
    sets = 1
  ): number[] {
    const spec = this.getScaleSpec(scaleType);
    return Array<number[]>(sets)
      .fill(spec)
      .reduce((acc, val) => [...acc, ...val])
      .reduce(
        (keys, nextKey) =>
          keys.length === 0
            ? [noteIndex, noteIndex + nextKey]
            : [...keys, keys[keys.length - 1] + nextKey],
        [] as number[]
      );
  }

  private static getScaleSpec(name: ScaleType): number[] {
    return Scale.scales[name];
  }
}
